### ISSUE TITLE
tf.keras.preprocessing.image_dataset_from_directory Error

### ISSUE THREAD

**Issue by daun4168** [tensorflow/tensorflow#48755]
TensorFlow version: 2.4.1
python version: 3.7.10
CUDA version: 11.2
CUDA Driver version: 465.19.01

Document of the function
  `tf.keras.preprocessing.image_dataset_from_directory`
the parameter `directory` is: 
`Directory where the data is located. If labels is "inferred", it should contain subdirectories, each containing images for a class. Otherwise, the directory structure is ignored.`

But the directory structure is not ignored.

I read a code of  `tf.keras.preprocessing.dataset_utils.index_directory`,
and I found that it always care about subdirs even if labels are feed by list already. 

Sample code: 
```python
train_ds = tf.keras.preprocessing.image_dataset_from_directory(
    image_path,
    labels=labels,
    label_mode="binary",
    validation_split=0.2,
    subset='training',
)

len(labels) == 202599
len(os.listdir(image_path)) == 202599  # Every file in image_path are JPG format.
```

It raises an error: 
 Expected the lengths of `labels` to match the number of files in the target directory. `len(labels)` is 202599 while we found 0 files in `image_path`.
